core/src/main/kotlin/xtdb/vector/IndirectMultiVectorReader.kt

```diff
@@ -0,0 +1,303 @@
+package xtdb.vector
+
+import clojure.lang.IFn
+import clojure.lang.Keyword
+import clojure.lang.RT
+import org.apache.arrow.memory.util.ArrowBufPointer
+import org.apache.arrow.memory.util.hash.ArrowBufHasher
+import org.apache.arrow.vector.NullVector
+import org.apache.arrow.vector.ValueVector
+import org.apache.arrow.vector.types.pojo.ArrowType
+import org.apache.arrow.vector.types.pojo.Field
+import xtdb.api.query.IKeyFn
+import xtdb.toLeg
+import xtdb.util.requiringResolve
+import java.nio.ByteBuffer
+import java.util.concurrent.ConcurrentHashMap
+
+class IndirectMultiVectorReader(
+    private val readers: List<IVectorReader?>,
+    private val readerIndirection: IVectorIndirection,
+    private val vectorIndirections: IVectorIndirection,
+) : IVectorReader {
+
+    private val fields: List<Field?> = readers.map { it?.field }
+    private val legReaders = ConcurrentHashMap<Keyword, IVectorReader>()
+    private val vectorField by lazy (LazyThreadSafetyMode.PUBLICATION){
+        MERGE_FIELDS.applyTo(RT.seq(fields.filterNotNull())) as Field
+    }
+
+    companion object {
+        private val MERGE_FIELDS: IFn = requiringResolve("xtdb.types/merge-fields")
+        private val VEC_TO_WRITER: IFn = requiringResolve("xtdb.vector.writer/->writer")
+    }
+
+    init {
+        assert(readers.any { it != null })
+    }
+
+    private fun unsupported(): RuntimeException {
+        throw UnsupportedOperationException("IndirectMultiVectoReader")
+    }
+
+    private fun safeReader(idx: Int): IVectorReader {
+        return readers[readerIndirection[idx]] ?: throw unsupported()
+    }
+
+    override fun valueCount(): Int {
+        return readerIndirection.valueCount()
+    }
+
+    override fun getName(): String {
+        return readers.filterNotNull().first().name
+    }
+
+    override fun withName(colName: String): IVectorReader {
+        return RenamedVectorReader(this, colName)
+    }
+
+    override fun getField(): Field {
+        return vectorField
+    }
+
+    override fun hashCode(idx: Int, hasher: ArrowBufHasher): Int {
+        return safeReader(idx).hashCode(vectorIndirections[idx], hasher)
+    }
+
+    override fun isNull(idx: Int): Boolean {
+        val readerIdx = readerIndirection[idx]
+        return readerIdx < 0 || readers[readerIdx] == null || readers[readerIdx]!!.isNull(vectorIndirections[idx])
+    }
+
+    override fun getBoolean(idx: Int): Boolean {
+        return safeReader(idx).getBoolean(vectorIndirections[idx])
+    }
```

Comment:
when it's a longer line, can still write it as an expression:
(IntelliJ will offer this as a refactoring)

```suggestion
    override fun getBoolean(idx: Int) =
        safeReader(idx).getBoolean(vectorIndirections[idx])
```